### PR TITLE
Medicine class

### PR DESCRIPTION
**Pull request by timetraveller-123**
Make medicine class
Change all string to medicine objects
Make addorder, add, edit check for invalid medicine and convert short form to medicine name
Check for invalid medicine in orders and allergies when loading storage
fix tests

### CODE REVIEW

**Comment by LinWanLeii** [comment]
Run coverage and try to at least git 75% coverage thanks

File: src/main/java/seedu/address/model/medicine/Medicine.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        checkArgument(isValidMedicineName(medicineName), MESSAGE_MEDICINE_NAME_CONSTRAINTS);
+        this.medicineName = medicineName;
+        this.shortForm = shorfForm;
+    }
+
+    public String getMedicineName() {
+        return this.medicineName;
+    }
+
+    public String getShorfForm() {
```
Comment (HEEaZ) [inline src/main/java/seedu/address/model/medicine/Medicine.java:51]: Typo in method name

File: src/main/java/seedu/address/model/medicine/Medicine.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        return this.shortForm;
+    }
+
+
+    /**
+     * Returns true if both medicines have the same medicine name.
+     * This defines a weaker notion of equality between two medicines.
+     */
+    public boolean isSameMedicine(Medicine m) {
```
Comment (HEEaZ) [inline src/main/java/seedu/address/model/medicine/Medicine.java:60]: This method can be simplified.

Use a guard clause to check if null:
After checking m == this, do
if (m == null) {
  return false;
}

Shortform equivalence is debatably unnecessary - but you can choose to leave it in, perhaps as 'defensive coding'

Perhaps there's some room to combine some of the if clauses

File: src/main/java/seedu/address/storage/JsonAdaptedOrder.java [hunk 3]
```diff
@@ -21,20 +28,22 @@ public class JsonAdaptedOrder {
...
     private final JsonAdaptedStatus orderStatus;
     /**
-     * Constructs a {@code JsonAdaptedOrder} with the given person details.
+     * Constructs a {@code JsonAdaptedOrder} with the given order details.
      */
     @JsonCreator
     public JsonAdaptedOrder(@JsonProperty("orderNumber") String orderNumber,
                             @JsonProperty("person") JsonAdaptedPerson person,
-                            @JsonProperty("medicineName") String medicineName,
+                            @JsonProperty("medicines") List<String> medicines,
```
Comment (HEEaZ) [inline src/main/java/seedu/address/storage/JsonAdaptedOrder.java:40]: This should be changed to 'List\<JsonAdaptedMedicine\>'

File: src/main/java/seedu/address/storage/JsonAdaptedOrder.java [hunk 4]
```diff
@@ -44,7 +53,7 @@ public JsonAdaptedOrder(@JsonProperty("orderNumber") String orderNumber,
     public JsonAdaptedOrder(Order order) {
         this.orderNumber = order.getOrderNumber().value;
         this.person = new JsonAdaptedPerson(order.getPerson());
-        this.medicineName = order.getMedicineName();
+        medicines.addAll(order.getMedicines().stream().map(Medicine::getMedicineName).collect(Collectors.toList()));
```
Comment (HEEaZ) [inline src/main/java/seedu/address/storage/JsonAdaptedOrder.java:56]: From JsonAdaptedPerson:        
```
tags.addAll(source.getTags().stream()
                .map(JsonAdaptedTag::new)
                .collect(Collectors.toList()));
```

Can employ similar method to get list of Medicine objects

File: src/main/java/seedu/address/storage/JsonAdaptedOrder.java [hunk 5]
```diff
@@ -21,20 +28,22 @@ public class JsonAdaptedOrder {
 
     private final JsonAdaptedPerson person;
 
-    private final String medicineName;
+    private final List<String> medicines = new ArrayList<>();
```
Comment (HEEaZ) [inline src/main/java/seedu/address/storage/JsonAdaptedOrder.java:31]: medicines here should be List\<JsonAdaptedMedicine\>

**Review by HEEaZ — CHANGES_REQUESTED** [review]
Summary:
1. Typo - quick fix
2. Storage files require changes: JsonAdaptedOrder, JsonAdaptedAllergy

Medicines in these classes should be JsonAdaptedMedicine instead of strings

**Review by HEEaZ — APPROVED** [review]
lgtm

**Review by LinWanLeii — APPROVED** [review]
LGTM

**Pull request merged**